go-controller/pkg/ovn/policy_retry.go

```diff
@@ -0,0 +1,186 @@
+package ovn
+
+import (
+	"fmt"
+	knet "k8s.io/api/networking/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"math/rand"
+	"time"
+
+	"k8s.io/klog/v2"
+)
+
+// iterateRetryNetworkPolicies checks if any outstanding NetworkPolicies exist
+// then tries to re-add them if so
+// updateAll forces all policies to be attempted to be retried regardless
+func (oc *Controller) iterateRetryNetworkPolicies(updateAll bool) {
+	oc.retryNetPolLock.Lock()
+	defer oc.retryNetPolLock.Unlock()
+	now := time.Now()
+	for namespacedName, npEntry := range oc.retryNetPolices {
+		if npEntry.ignore {
+			continue
+		}
+
+		// check if we need to create
+		var policyToCreate *knet.NetworkPolicy
+		if npEntry.newPolicy != nil {
+			// get the latest version of the new policy from the informer, if it doesn't exist we are not going to
+			// create the new policy
+			np, err := oc.watchFactory.GetNetworkPolicy(npEntry.newPolicy.Namespace, npEntry.newPolicy.Name)
+			if err != nil && errors.IsNotFound(err) {
+				klog.Infof("%s policy not found in the informers cache, not going to retry policy create", namespacedName)
+				npEntry.newPolicy = nil
+			} else {
+				policyToCreate = np
+			}
+		}
+
+		npEntry.backoffSec = npEntry.backoffSec * 2
+		if npEntry.backoffSec > 60 {
+			npEntry.backoffSec = 60
+		}
+		backoff := (npEntry.backoffSec * time.Second) + (time.Duration(rand.Intn(500)) * time.Millisecond)
+		npTimer := npEntry.timeStamp.Add(backoff)
+		if updateAll || now.After(npTimer) {
+			klog.Infof("Network Policy Retry: %s retry network policy setup", namespacedName)
+
+			// check if we need to delete anything
+			if npEntry.oldPolicy != nil {
+				klog.Infof("Network Policy Retry: Removing old policy for %s", namespacedName)
+				if err := oc.deleteNetworkPolicy(npEntry.oldPolicy, npEntry.np); err != nil {
+					klog.Infof("Network Policy Retry delete failed for %s, will try again later: %v",
+						namespacedName, err)
+					npEntry.timeStamp = time.Now()
+					continue
+				}
+				// successfully cleaned up old policy, remove it from the retry cache
+				npEntry.oldPolicy = nil
+			}
+
+			// create new policy if needed
+			if npEntry.newPolicy != nil && policyToCreate != nil {
```

Comment:
isn't this check for `npEntry.newPolicy != nil` redundant since we do it way before in L27 and already have policyToCreate be the variable that tells us if we need to retry add?